### PR TITLE
made fixes for translations, added support for levels, improved print…

### PR DESCRIPTION
**Pull request by m-kovalsky**
…outs

### CODE REVIEW

File: src/sempy_labs/_translations.py [hunk 1]
```diff
@@ -101,6 +101,20 @@ def translate_semantic_model(
...
+
+            for o in tom.all_levels():
+                new_data = {
+                    "Object Type": "Level",
+                    "Name": o.Name,
+                    "TName": o.Name,
+                    "Description": o.Description,
+                    "TDescription": o.Description,
+                    "Display Folder": None,
+                    "TDisplay Folder": None,
```
Comment (eisber) [inline src/sempy_labs/_translations.py:113]: what does T stand for? translation?

File: src/sempy_labs/_translations.py [hunk 2]
```diff
@@ -101,6 +101,20 @@ def translate_semantic_model(
...
+                }
+                dfPrep = pd.concat(
```
Comment (eisber) [inline src/sempy_labs/_translations.py:115]: nit: df_prep

File: src/sempy_labs/_translations.py [hunk 3]
```diff
@@ -330,6 +355,33 @@ def translate_semantic_model(
...
                                         )
+                                for lev in h.Levels:
+                                    if clm == "Name":
+                                        df_filt = df_panda[
+                                            (df_panda["Object Type"] == "Level")
+                                            & (df_panda["Name"] == lev.Name)
+                                        ]
+                                        if len(df_filt) == 1:
+                                            tr = df_filt["translation"].str[i].iloc[0]
+                                            tom.set_translation(
```
Comment (eisber) [inline src/sempy_labs/_translations.py:366]: can you refactor into functions/methods? the code looks quite similar. also the long function makes it hard to follow

**Review by eisber — CHANGES_REQUESTED** [review]

File: src/sempy_labs/_translations.py (hunk 1 above)
Comment (m-kovalsky) [reply src/sempy_labs/_translations.py:113]: @eisber there are 2 columns for each (TName and Name) because of the exclude_characters functionality. Since it may change the object name there are 2 versions of each (one to maintain the original name and one to have the updated name (i.e 'Internet_Sales' and 'Internet Sales'). I can probably make this cleaner.

File: src/sempy_labs/_translations.py (hunk 2 above)
Comment (m-kovalsky) [reply src/sempy_labs/_translations.py]: updated

File: src/sempy_labs/_translations.py (hunk 3 above)
Comment (m-kovalsky) [reply src/sempy_labs/_translations.py]: made 2 subfunctions within this function to greatly simplify the code as recommended.

File: src/sempy_labs/tom/_model.py [hunk 4]
```diff
@@ -1848,7 +1895,7 @@ def hybrid_tables(self):
                 if any(p.Mode == TOM.ModeType.DirectQuery for p in t.Partitions):
                     yield t
 
-    def date_tables(self):
+    def all_date_tables(self):
```
Comment (eisber) [inline src/sempy_labs/tom/_model.py:1898]: if you rename methods (which breaks the API) you need to increase the major version

File: src/sempy_labs/migration/_refresh_calc_tables.py [hunk 5]
```diff
@@ -121,9 +122,11 @@ def refresh_calc_tables(dataset: str, workspace: Optional[str] = None):
                                 f"{icons.green_dot} Calculated table '{tName}' has been refreshed as the '{delta_table_name.lower()}' table in the lakehouse."
                             )
                         except Exception as e:
```
Comment (eisber) [inline src/sempy_labs/migration/_refresh_calc_tables.py:124]: for a separate PR, we should do a round of exception reviewing. I think there's still a bunch of cases, where you can check on a condition before hand and avoid just catching the exception

**Review by eisber — APPROVED** [review]
